src/tests/unittests/d3d12/CopySplit.cpp

```diff
@@ -0,0 +1,375 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+
+#include "backend/d3d12/d3d12_platform.h"
+#include "backend/d3d12/TextureCopySplitter.h"
+#include "common/Assert.h"
+#include "common/Constants.h"
+#include "common/Math.h"
+
+using namespace backend::d3d12::TextureCopy;
+
+namespace {
+
+    struct TextureSpec {
+        uint32_t x;
+        uint32_t y;
+        uint32_t z;
+        uint32_t width;
+        uint32_t height;
+        uint32_t depth;
+        uint32_t texelSize;
+    };
+
+    struct BufferSpec {
+        uint32_t offset;
+        uint32_t rowPitch;
+    };
+
+    // Check that each copy region fits inside the buffer footprint
+    void ValidateFootprints(const CopySplit& copySplit) {
+        for (uint32_t i = 0; i < copySplit.count; ++i) {
+            const CopyInfo& copy = copySplit.copies[i];
+            ASSERT_LE(copy.buffer.x + copy.texture.width, copy.buffer.width);
+            ASSERT_LE(copy.buffer.y + copy.texture.height, copy.buffer.height);
+            ASSERT_LE(copy.buffer.z + copy.texture.depth, copy.buffer.depth);
+        }
+    }
+
+    // Check that the offset is aligned
+    void ValidateOffset(const CopySplit& copySplit) {
+        ASSERT_TRUE(Align(copySplit.offset, D3D12_TEXTURE_DATA_PLACEMENT_ALIGNMENT) == copySplit.offset);
+    }
+
+    bool RangesOverlap(uint32_t minA, uint32_t maxA, uint32_t minB, uint32_t maxB) {
+        return (minA < minB && minB >= maxA) || (minB < minA && minA >= maxB);
+    }
+
+    // Check that no pair of copy regions intersect each other
+    void ValidateDisjoint(const CopySplit& copySplit) {
+        for (uint32_t i = 0; i < copySplit.count; ++i) {
+            const CopyInfo& a = copySplit.copies[i];
+            for (uint32_t j = i + 1; j < copySplit.count; ++j) {
+                const CopyInfo& b = copySplit.copies[j];
+                bool overlapX = RangesOverlap(a.texture.x, a.texture.x + a.texture.width, b.texture.x, b.texture.x + b.texture.width);
+                bool overlapY = RangesOverlap(a.texture.y, a.texture.y + a.texture.height, b.texture.y, b.texture.y + b.texture.height);
+                bool overlapZ = RangesOverlap(a.texture.z, a.texture.z + a.texture.depth, b.texture.z, b.texture.z + b.texture.depth);
+                ASSERT_TRUE(!overlapX || !overlapY || !overlapZ);
+            }
+        }
+    }
+
+    // Check that the union of the copy regions exactly covers the texture region
+    void ValidateTextureCoverage(const TextureSpec& textureSpec, const CopySplit& copySplit) {
+        ASSERT_TRUE(copySplit.count > 0);
```

Comment:
This doesn't check that all pixels are covered, just that they are all in bounds. Can we have two checks that validate that both the regions are in bounds of the texture, and another one that checks the number of pixel copied matches what we expect?